config.go

```diff
@@ -73,6 +73,10 @@ type Config struct {
 	ServerName string
 
 	LoggerFactory logging.LoggerFactory
+
+	// ConnectTimeoutMs is the timeout threshold for new connection handshakes
+	// to complete (default is no timeout)
```

Comment:
I don't think the timeout should be no timeout by default. I realise that's the current behaviour but if we're going to introduce a way to control it I'd also prefer we set a value, 30s for example. No timeout feels like just leaving a gun for someone to shoot themselves with. If people really want no timeout they can explicitly set the timeout to 0.

Reply:
That works for me but how do you think we should default it? Golang by default initializes it to zero so its difficult to tell if its explicitly set to zero or implicitly initialized to zero. I could turn it into a pointer with nil as default or move config to a constructor based system

Reply:
I don't think the -1/0 is what we should do. So far in all Go libraries I've seen dealing with networks, the value 0 w.r.t timeouts has meant no timeout. We should stick to that, -1 would be surprising.

I would be inclined to make it a pointer but it's annoying when using the library since you now need a `IntP` helper or something to give you an integer pointer.

@Sean-Der @igolaizola thoughts?

Reply:
I'm thinking `*time.Duration` or `*int` might be the least of all evils/road-blocky. We can have a little function for people to use:

```
func ConnectTimeoutOption(time.Duration) *time.Duration
```

Any of the other options (variadic functions, constructors) are larger breaking changes

Reply:
Ah yes, that works for me.

Reply:
Okay updated! :+1:

Reply:
In my opinion no timeout by default doesn't seems like a bad idea.
- A default value of 30 sec might not fit every user.
- Using a `time.Duration` is much cleaner than `*time.Duration`.

Some examples that use a `Timeout time.Duration` property and "no timeout" as default:
- `http.Client` of the standard library https://golang.org/pkg/net/http/#Client 
- `http2.Server` https://godoc.org/golang.org/x/net/http2#Server

Reply:
Fantastic feature @RyanGordon, we are going to do a little bike shedding. But I think this is for sure a merge!

----
Oh nice, thanks for the data @igolaizola. I would say copy stdlib w/e we can. Feels like the source of truth in all cases :)

Does anyone know why `crypto/tls` doesn't have a timeout? 

I have never had a case where I needed timeouts with crypto/tls but it looks like Deadline is the way to do it. Should we copy them and use `Deadline` to cover this (instead of timeout?) then the deadline would apply to every read as well?

Reply:
The no timeout as a default is dangerous. It makes this library a danger to use in production if people aren't aware of it. http.Server, http.Client etc not having a timeout is b/c nobody considered it before it landed in Go 1 and they can't change it now. It's why not safe to use http.Server and why there's articles like https://blog.cloudflare.com/exposing-go-on-the-internet/ that try to teach people to set timeouts.

Reply:
Thanks for the push back @daenney, you have convinced me :) 

My vote is 30 seconds as well, we should also update the examples to maybe be 30 seconds explicitly so people expect it?

Reply:
@daenney you've also convinced me. Thanks for the cloudflare's article link btw, it's very interesting.

The only way to avoid `*time.Duration` will be using `MaxInt` for "no timeout" and reinterpreting the zero value as default, but it might be confusing as you've already said.

Reply:
Kept at 30 seconds and updated examples :+1: